### PR TITLE
[Native DSL] Quack-based cuteDSL RMSNorm

### PR DESCRIPTION
**Pull request by slayton58**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #182108

Summary:

Add fused rmsnorm fwd/bwd implementations from quack, written in
cutedsl. Rely on vendored quack.

Add testing to OpInfo for composability with other PyT systems, and
test_nn to match the existing fused norm tests.

Test Plan:

```
  python -m pytest -v \
    test/test_nn.py -k rmsnorm \
    test/python_native/ \
    test/test_ops.py -k 'rms_norm' \
    test/inductor/test_torchinductor_opinfo.py -k 'rms_norm'

  # Then separately, since OPINFO_RESTRICT_TO_DSL filters op_db globally:
  OPINFO_RESTRICT_TO_DSL=cutedsl python -m pytest -v \
    test/test_ops.py -k 'rms_norm and cutedsl' \
    test/inductor/test_torchinductor_opinfo.py -k 'rms_norm and cutedsl' \
    test/export/test_export_opinfo.py -k 'rms_norm and cutedsl'
```

Authors:

Originally from #178326

Co-authored-by: @AaronWang04

Signed-off-by: Simon Layton <simonlayton@meta.com>
Signed-off-by: Simon Layton <simonlayton@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_native/ops/norm/norms.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    M = input.numel() // N
+    # quack's kernel requires a contiguous 2-D input with trailing stride 1;
+    # a plain `.reshape(M, N)` can return a non-contiguous view.
+    x = _reshape_2d(input, M, N)
+
+    out = torch.empty_like(x)
+    rstd = torch.empty(M, device=x.device, dtype=torch.float32)
+
+    if weight is not None:
+        weight = weight.reshape(N).contiguous()
```
Comment (drisspg) [inline torch/_native/ops/norm/norms.py:58]: hmm did we measure the impact on perf for this path?

File: torch/_native/ops/norm/norms.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+import torch
+
+
+# quack imports `cutlass`, which is only installed on CUDA-enabled x86 Linux
+# builds. Keep the import lazy so `torch._native.ops.norm.norms` stays
+# importable on CPU-only builds (test_public_bindings /
+# test_circular_dependencies walk every module under `torch.`).
+def _torch2cute(t: torch.Tensor | None):  # type: ignore[no-untyped-def]
+    from torch._vendor.quack import cute_dsl_utils as _cute_dsl_utils
```
Comment (drisspg) [inline torch/_native/ops/norm/norms.py:20]: should we define / maybe reexport in our cute dsl utils

File: torch/_native/ops/norm/rmsnorm_impl.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+
+def _fused_rms_norm_impl(
+    input: torch.Tensor,
+    normalized_shape: list[int],
+    weight: torch.Tensor | None,
+    eps: float | None,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    if eps is None:
+        eps = torch.finfo(input.dtype).eps
```
Comment (drisspg) [inline torch/_native/ops/norm/rmsnorm_impl.py]: the cuda impl looks to be creating Eps off of the accumulator type not the input dtype so

File: torch/_native/ops/norm/rmsnorm_impl.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        return False
+    # The override reshapes + makes contiguous, which materializes a COW input.
+    # Match the bmm_outer_product cond (triton_impl.py:46) and fall through to
+    # aten so composite-compliance tests don't flag spurious materialization.
+    is_cow = torch._C._is_cow_tensor  # pyrefly: ignore[missing-attribute]
+    if is_cow(input):
+        return False
+    if weight is not None and is_cow(weight):
+        return False
+    return True
```
Comment (drisspg) [inline torch/_native/ops/norm/rmsnorm_impl.py:73]: out of  curiosity: I think the CUDA impl does some input validation. We are doing all the same input validation for the approved paths, right?

File: torch/_native/ops/norm/norms.py (hunk 1 above)
Comment (slayton58) [reply torch/_native/ops/norm/norms.py:58]: Forcing weight to be contiguous? No, I don't think that's been measured, I think, realistically, we'd probably bail on the override in that case (not that I think it's a common one tbh)

File: torch/_native/ops/norm/norms.py (hunk 2 above)
Comment (slayton58) [reply torch/_native/ops/norm/norms.py:20]: Actually surprised this isn't in quack (or perhaps it is and needs to be called here), but shared functions across a backend, but not through quack is a bit of a blind spot right now (it came up in the `scatter_add` review as well)

File: torch/_native/ops/norm/rmsnorm_impl.py (hunk 4 above)
Comment (slayton58) [reply torch/_native/ops/norm/rmsnorm_impl.py:73]: That needs to be checked, but is another common thing from the `scatter_add` review -- override paths need to be very careful with their input validation

File: torch/_native/ops/norm/rmsnorm_impl.py (hunk 3 above)
Comment (slayton58) [reply torch/_native/ops/norm/rmsnorm_impl.py]: Fixed

File: torch/_native/ops/norm/rmsnorm_impl.py (hunk 4 above)
Comment (slayton58) [reply torch/_native/ops/norm/rmsnorm_impl.py:73]: Updated so the override won't take cases the aten op won't -- it'll fall back and let the existing error checking in aten fire.

File: torch/_native/ops/norm/norms.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    if weight is not None:
+        weight = weight.reshape(N).contiguous()
+
+    # quack treats `weight_dtype` (whether weight is present in the kernel)
+    # and `has_dw_partial` (whether to accumulate dw) as independent flags.
+    # dx's formula depends on weight, so we always hand weight to the kernel
+    # when the user supplied one; dw_mask only gates the dw_partial buffer.
+    dw_partial: torch.Tensor | None = None
+    if weight is not None and dw_mask:
+        dw_partial = torch.empty(sm_count, N, device=x.device, dtype=torch.float32)
```
Comment (drisspg) [inline torch/_native/ops/norm/norms.py:119]: Curious: why output_mask for only 1 of em? is that just what quack supports?

File: test/test_nn.py [hunk 6]
```diff
@@ -14515,6 +14515,290 @@ def __init__(self) -> None:
...
+        if output_mask[0]:
+            self.assertIsNotNone(gx)
+            self.assertEqual(gx, gx_ref, atol=1e-1, rtol=0)
+        else:
+            self.assertIsNone(gx)
+        if output_mask[1]:
+            self.assertIsNotNone(gw)
+            self.assertEqual(gw, gw_ref, atol=1e-1, rtol=0)
+        else:
+            self.assertIsNone(gw)
```
Comment (drisspg) [inline test/test_nn.py:14817]: nit: should we have 1 test that acutally uses the profiler to ensure the right kernel is called?

File: torch/_native/ops/norm/norms.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    )
+    # compile order: (x, weight, dout, dres_out, rstd, dx, dw_partial, dres, db_partial, sm_count)
+    kernel(x, weight, dout, None, rstd_flat, dx, dw_partial, None, None, sm_count)
+
+    dx = dx.reshape(input.shape)
+    # `dw_partial is not None` implies `weight is not None` (invariant from
+    # allocation above), but pyrefly can't see the dependency.
+    if dw_partial is not None:
+        weight_dtype_for_sum = weight.dtype  # pyrefly: ignore[missing-attribute]
+        dw = dw_partial.sum(dim=0, dtype=weight_dtype_for_sum).reshape(normalized_shape)
```
Comment (drisspg) [inline torch/_native/ops/norm/norms.py]: Codex has convinced me that this will cast the fp32 dw_partial to the weight dtype before reduction and then sum intermediates in fp32 again but by then you have already gone though a cas

File: test/test_nn.py (hunk 6 above)
Comment (slayton58) [reply test/test_nn.py:14817]: Funnily enough there used to be that, and I moved to testing the conditional, and then testing the kernel on cases that the conditional was shown to accept.

The case that a profile would catch is a bit of a nasty one - where the conditional would evaluate true, but the kernel wouldn't fire -- I think this could happen if some overrides are unexpectedly disabled, or (more likely I think) when there's multiple overrides with overlapping conditionals, and a different override is called instead.

This is a bunch of words to say I think there's value in this, but I'm not 100% sure how this should look -- per-override-op tests? Centralized test where each override registers a set of `(test-fn, kernel_name)` pairs, and we have a single "are the correct overrides firing for each of these cases" test?

File: torch/_native/ops/norm/norms.py (hunk 5 above)
Comment (slayton58) [reply torch/_native/ops/norm/norms.py:119]: It looks kinda like quack doesn't support -- there's db_partial, but there's no docstring, and dx is explicitly called out as the input gradient, so it can't relate to that

File: torch/_native/ops/norm/norms.py (hunk 7 above)
Comment (slayton58) [reply torch/_native/ops/norm/norms.py]: Is the concern here where `weight` is a lower precision than `dw_partial`?

File: test/test_nn.py (hunk 6 above)
Comment (drisspg) [reply test/test_nn.py:14817]: yeah I was more thinking of like a reallly dumb sanity check but I agree that perhaps this belt and suspenders

File: torch/_native/ops/norm/norms.py (hunk 5 above)
Comment (drisspg) [reply torch/_native/ops/norm/norms.py:119]: yeah I buy it

File: torch/_native/ops/norm/norms.py (hunk 7 above)
Comment (drisspg) [reply torch/_native/ops/norm/norms.py]: Yeah exactly; you will end up 

partials (fp32) -> cast(partials to weight dtype) -> sum(in fp32) -> cast to (weight dtype)

instead of 
partials(fp32) -> sum(fp32) -> cast(acc to weight_dtype)

Comment (slayton58) [reply torch/_native/ops/norm/norms.py]: Gotcha - means:
```
dw = dw_partial.sum(dim=0).reshape(normalized_shape).to(weight_dtype_for_sum)
```
is what we're after

**Comment by drisspg** [comment]
@claude Please do a deep review focused on subtle correctness bugs and regressions that are easy for human reviewers to miss. Prioritize findings over summary. I dont care about anything that isnt a real bug drop all the noise but that, if you there isnt anything to report just say so with a small comment

**Review by drisspg — APPROVED** [review]
